File: knowledgebase-sdk/src/main/java/ru/usedesk/knowledgebase_sdk/internal/data/repository/ApiRepository.java
package ru.usedesk.knowledgebase_sdk.internal.data.repository;

import androidx.annotation.NonNull;

import java.util.Arrays;
import java.util.List;

import javax.inject.Inject;

import ru.usedesk.common_sdk.external.entity.exceptions.UsedeskDataNotFoundException;
import ru.usedesk.common_sdk.external.entity.exceptions.UsedeskHttpException;
import ru.usedesk.knowledgebase_sdk.external.entity.UsedeskArticleBody;
import ru.usedesk.knowledgebase_sdk.external.entity.UsedeskArticleInfo;
import ru.usedesk.knowledgebase_sdk.external.entity.UsedeskCategory;
import ru.usedesk.knowledgebase_sdk.external.entity.UsedeskSection;
import ru.usedesk.knowledgebase_sdk.external.entity.UsedeskSearchQuery;
import ru.usedesk.knowledgebase_sdk.internal.data.framework.retrofit.IApiLoader;

public class ApiRepository implements IKnowledgeBaseRepository {

    private IApiLoader apiLoader;

    private List<UsedeskSection> sectionList;

    @Inject
    ApiRepository(IApiLoader apiLoader) {
        this.apiLoader = apiLoader;
    }

    @NonNull
    @Override
    public List<UsedeskSection> getSections(@NonNull String id, @NonNull String token) throws UsedeskHttpException {
        if (sectionList == null) {
            sectionList = Arrays.asList(apiLoader.getSections(id, token));
        }
        return sectionList;
    }

    @NonNull
    @Override
    public UsedeskArticleBody getArticleBody(@NonNull String id, @NonNull String token, long articleId) throws UsedeskHttpException {
        return apiLoader.getArticle(id, Long.toString(articleId), token);
    }

    @NonNull
    @Override
    public List<UsedeskArticleBody> getArticles(@NonNull String accountId, @NonNull String token,
                                                @NonNull UsedeskSearchQuery searchQuery) throws UsedeskHttpException {
        return apiLoader.getArticles(accountId, token, searchQuery);
    }

    @NonNull
    @Override
    public List<UsedeskCategory> getCategories(@NonNull String id, @NonNull String token, long sectionId)
            throws UsedeskHttpException, UsedeskDataNotFoundException {
        if (sectionList == null) {
            getSections(id, token);
        }
        return Arrays.asList(getCategories(sectionId));
    }

    @NonNull
    @Override
    public List<UsedeskArticleInfo> getArticles(@NonNull String id, @NonNull String token, long categoryId)
            throws UsedeskHttpException, UsedeskDataNotFoundException {
        if (sectionList == null) {
            getSections(id, token);
        }
        return Arrays.asList(getArticles(categoryId));
    }

    @Override
    public void addViews(@NonNull String accountId, @NonNull String token, long articleId)
            throws UsedeskHttpException, UsedeskDataNotFoundException {
        int views = apiLoader.addViews(accountId, token, articleId, 1);

        getArticleBody(accountId, token, articleId).setViews(views);
        getArticleInfo(articleId).setViews(views);
    }

    private UsedeskArticleInfo getArticleInfo(long articleId) throws UsedeskDataNotFoundException {
        for (UsedeskSection section : sectionList) {
            for (UsedeskCategory category : section.getCategories())
                for (UsedeskArticleInfo articleInfo : category.getArticles()) {
                    if (articleInfo.getId() == articleId) {
                        return articleInfo;
                    }
                }
        }
        throw new UsedeskDataNotFoundException();
    }

    @NonNull
    private UsedeskArticleInfo[] getArticles(long categoryId) throws UsedeskDataNotFoundException {
        for (UsedeskSection section : sectionList) {
            for (UsedeskCategory category : section.getCategories())
                if (category.getId() == categoryId) {
                    return category.getArticles();
                }
        }
        throw new UsedeskDataNotFoundException("UsedeskCategory with id(" + categoryId + ")");
    }

    @NonNull
    private UsedeskCategory[] getCategories(long sectionId) throws UsedeskDataNotFoundException {
        for (UsedeskSection section : sectionList) {
            if (section.getId() == sectionId) {
                return section.getCategories();
            }
        }
        throw new UsedeskDataNotFoundException("UsedeskSection with id(" + sectionId + ")");
    }
}

File: knowledgebase-sdk/src/main/java/ru/usedesk/knowledgebase_sdk/internal/data/repository/ApiRepository.kt
package ru.usedesk.knowledgebase_sdk.internal.data.repository

import ru.usedesk.common_sdk.external.entity.exceptions.UsedeskDataNotFoundException
import ru.usedesk.common_sdk.external.entity.exceptions.UsedeskHttpException
import ru.usedesk.knowledgebase_sdk.external.entity.*
import ru.usedesk.knowledgebase_sdk.internal.data.framework.retrofit.IApiLoader
import toothpick.InjectConstructor

@InjectConstructor
class ApiRepository(
        private val apiLoader: IApiLoader
) : IKnowledgeBaseRepository {

    private var sectionList: List<UsedeskSection>? = null

    @Throws(UsedeskHttpException::class)
    override fun getSections(accountId: String, token: String): List<UsedeskSection> {
        if (sectionList == null) {
            sectionList = apiLoader.getSections(accountId, token).toList()
        }
        return sectionList!!
    }

    @Throws(UsedeskHttpException::class)
    override fun getArticleBody(accountId: String, token: String, articleId: Long): UsedeskArticleBody {
        return apiLoader.getArticle(accountId, articleId.toString(), token)
    }

    @Throws(UsedeskHttpException::class)
    override fun getArticles(accountId: String,
                             token: String,
                             searchQuery: UsedeskSearchQuery): List<UsedeskArticleBody> {
        return apiLoader.getArticles(accountId, token, searchQuery)
    }

    @Throws(UsedeskHttpException::class, UsedeskDataNotFoundException::class)
    override fun getCategories(accountId: String, token: String, sectionId: Long): List<UsedeskCategory> {
        if (sectionList == null) {
            getSections(accountId, token)
        }
        return getCategories(sectionId).toList()
    }

    @Throws(UsedeskHttpException::class, UsedeskDataNotFoundException::class)
    override fun getArticles(accountId: String, token: String, categoryId: Long): List<UsedeskArticleInfo> {
        if (sectionList == null) {
            getSections(accountId, token)
        }
        return getArticles(categoryId).toList()
    }

    @Throws(UsedeskHttpException::class, UsedeskDataNotFoundException::class)
    override fun addViews(accountId: String, token: String, articleId: Long) {
        val views = apiLoader.addViews(accountId, token, articleId, 1)
        getArticleBody(accountId, token, articleId).views = views
        getArticleInfo(articleId).views = views
    }

    @Throws(UsedeskDataNotFoundException::class)
    private fun getArticleInfo(articleId: Long): UsedeskArticleInfo {
        return sectionList?.asSequence()?.mapNotNull { section ->
            section.categories
        }?.flatMap { categories ->
            categories.asSequence()
        }?.mapNotNull { category ->
            category.articles
        }?.flatMap { articles ->
            articles.asSequence()
        }?.first { article ->
            article.id == articleId
        } ?: throw UsedeskDataNotFoundException("UsedeskArticleInfo with id($articleId)")
    }

    @Throws(UsedeskDataNotFoundException::class)
    private fun getArticles(categoryId: Long): Array<UsedeskArticleInfo> {
        return sectionList?.mapNotNull { section ->
            section.categories
        }?.flatMap { categories ->
            categories.asSequence()
        }?.firstOrNull { category ->
            category.id == categoryId
        }?.articles ?: throw UsedeskDataNotFoundException("UsedeskCategory with id($categoryId)")
    }

    @Throws(UsedeskDataNotFoundException::class)
    private fun getCategories(sectionId: Long): Array<UsedeskCategory> {
        return sectionList?.first { section ->
            section.id == sectionId
        }?.categories ?: throw UsedeskDataNotFoundException("UsedeskSection with id($sectionId)")
    }
}